lib-extra/src/main/java/com/diffplug/spotless/extra/npm/PrettierFormatterStep.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright 2016 DiffPlug
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.diffplug.spotless.extra.npm;
+
+import static java.util.Arrays.asList;
+import static java.util.Objects.requireNonNull;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.Map;
+import java.util.TreeMap;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import com.diffplug.spotless.FormatterFunc;
+import com.diffplug.spotless.FormatterStep;
+import com.diffplug.spotless.Provisioner;
+import com.diffplug.spotless.ThrowingEx;
+
+public class PrettierFormatterStep {
+
+	public static final String NAME = "prettier-format";
+
+	public static FormatterStep create(Provisioner provisioner, File buildDir, @Nullable File npm, PrettierConfig prettierConfig) {
+		requireNonNull(provisioner);
+		requireNonNull(buildDir);
+		return FormatterStep.createLazy(NAME,
+				() -> new State(NAME, provisioner, buildDir, npm, prettierConfig),
+				State::createFormatterFunc);
+	}
+
+	public static class State extends NpmFormatterStepStateBase implements Serializable {
+
+		private static final long serialVersionUID = -3811104513825329168L;
+		private final PrettierConfig prettierConfig;
+
+		State(String stepName, Provisioner provisioner, File buildDir, @Nullable File npm, PrettierConfig prettierConfig) throws IOException {
+			super(stepName,
+					provisioner,
+					new NpmConfig(
+							readFileFromClasspath(PrettierFormatterStep.class, "/com/diffplug/spotless/extra/npm/prettier-package.json"),
+							"prettier"),
+					buildDir,
+					npm);
+			this.prettierConfig = requireNonNull(prettierConfig);
+		}
+
+		@Override
+		@Nonnull
+		public FormatterFunc createFormatterFunc() {
+
+			try {
+				final NodeJSWrapper nodeJSWrapper = nodeJSWrapper();
+				final V8ObjectWrapper prettier = nodeJSWrapper.require(nodeModulePath());
+
+				@SuppressWarnings("unchecked")
+				final Map<String, Object>[] resolvedPrettierOptions = (Map<String, Object>[]) new Map[1];
+				if (this.prettierConfig.getPrettierConfigPath() != null) {
+					final Exception[] toThrow = new Exception[1];
+					try (
+							V8FunctionWrapper resolveConfigCallback = nodeJSWrapper.createNewFunction((receiver, parameters) -> {
```

Comment:
That's some wacky formatting here

Can I suggest you put some of this logic into outer functions?

Reply:
agreed! just pushed a change to extract some of the cluttering object creation statements